src/model/formats/foreflight-utils.ts

```diff
@@ -0,0 +1,93 @@
+import { ChecklistGroup_Category } from '../../../gen/ts/checklist';
+import { v4 as uuid } from 'uuid';
+import { ForeFlightFormatError } from './foreflight-format';
+
+export enum ForeFlightCategory {
+  Normal = 'normal',
+  Abnormal = 'abnormal',
+  Emergency = 'emergency',
+}
+
+enum KeyUsage {
+  Encrypt = 'encrypt',
+  Decrypt = 'decrypt',
+}
+
+export class ForeFlightUtils {
+  public static readonly FILE_EXTENSION = 'fmd';
+
+  public static readonly CONTAINER_TYPE = 'checklist';
+  public static readonly SCHEMA_VERSION = '1.0';
+
+  public static readonly ITEM_HEADER = 'comment';
+  public static readonly NOTE_INDENT = 1;
+
+  public static readonly OBJECT_ID_LENGTH = 32;
+
+  static readonly CIPHER_TYPE = 'AES-CBC';
+  static readonly CIPHER_BLOCK_SIZE = 16;
+  static readonly CIPHER_KEY = Buffer.from('81e06e41a93f3848', 'ascii');
+
+  static readonly encoder = new TextEncoder();
+  static readonly decoder = new TextDecoder();
+
+  private static getKey(keyUsage: KeyUsage): Promise<CryptoKey> {
+    return window.crypto.subtle.importKey(
+      'raw',
+      ForeFlightUtils.CIPHER_KEY,
+      { name: ForeFlightUtils.CIPHER_TYPE },
+      false,
+      [keyUsage],
+    );
+  }
+
+  public static async decrypt(stream: ArrayBuffer): Promise<string> {
+    const decrypted = await window.crypto.subtle.decrypt(
+      {
+        name: ForeFlightUtils.CIPHER_TYPE,
+        iv: new Uint8Array(stream.slice(0, ForeFlightUtils.CIPHER_BLOCK_SIZE)),
+      },
+      await ForeFlightUtils.getKey(KeyUsage.Decrypt),
+      new Uint8Array(stream.slice(ForeFlightUtils.CIPHER_BLOCK_SIZE)),
+    );
+    return ForeFlightUtils.decoder.decode(new Uint8Array(decrypted));
+  }
+
+  public static async encrypt(data: string): Promise<Blob> {
+    const iv = window.crypto.getRandomValues(new Uint8Array(ForeFlightUtils.CIPHER_BLOCK_SIZE));
+    return new Blob(
+      [
+        iv,
+        await window.crypto.subtle.encrypt(
+          { name: ForeFlightUtils.CIPHER_TYPE, iv: iv },
+          await ForeFlightUtils.getKey(KeyUsage.Encrypt),
+          ForeFlightUtils.encoder.encode(data),
+        ),
+      ],
+      { type: 'application/octet-stream' },
+    );
+  }
+
+  public static categoryToEFIS(category: string): ChecklistGroup_Category {
```

Comment:
Doesn't proto support defining the field as the enum directly and parsing from the strings, instead of needing constants + this function?

Reply:
Unfortunately, not to my knowledge, at least not according to this issue: https://github.com/protocolbuffers/protobuf/issues/8483. The documentation that I've found says that enums must be integer-based (https://protobuf.dev/programming-guides/proto3/#enum) and the generated code for `ChecklistGroup_Category` doesn't even include a reverse mapping.

I could try to do some reflection magic instead of a switch for the conversion though, and I would have done it like 10 years ago, but these days I feel like the complexity and type erasure is not worth it for 3 elements...

Please let me know if you have a better solution, maybe I'm still missing something.

Reply:
That's weird, 'cause when I export the protos in JSON format ("raw format" in the UI), it gives me output like this:

```
            {
              "prompt": "Fuel required",
              "expectation": "COMPUTED",
              "type": "ITEM_CHALLENGE_RESPONSE"
            },
```

which it also parses back just fine - so it is using the string of the enum in the JSON.

I just downloaded your code and took a stab at this - here's an (incomplete) patch. The key thing is parsing the JSON with the proto class itself. The caveat that I'm not sure there's a workaround for (and why the patch is incomplete, I didn't have time so far to dig further) is that the parser seems to be case-sensitive for the enum names :/

```
diff --git a/src/model/formats/foreflight-format.spec.ts b/src/model/formats/foreflight-format.spec.ts
index 75f4e21530...d56764ec4f 100644
--- a/src/model/formats/foreflight-format.spec.ts
+++ b/src/model/formats/foreflight-format.spec.ts
@@ -1,5 +1,4 @@
-import { ChecklistGroup_Category } from '../../../gen/ts/checklist';
-import { ForeFlightCategory, ForeFlightUtils } from './foreflight-utils';
+import { ForeFlightUtils } from './foreflight-utils';
 import { EXPECTED_FOREFLIGHT_CONTENTS } from './test-data';
 import { ForeFlightReader } from './foreflight-reader';
 import { loadFile } from './test-utils';
@@ -22,13 +21,6 @@
       expect(decrypted).toBe(checklistData);
     });

-    it('should map ForeFlight categories to EFIS', async () => {
-      expect(ForeFlightUtils.categoryToEFIS(ForeFlightCategory.Normal)).toBe(ChecklistGroup_Category.NORMAL);
-      expect(() => {
-        ForeFlightUtils.categoryToEFIS('undefined');
-      }).toThrowError("ForeFlight: unknown category 'undefined'");
-    });
-
     it('should determine correct checklist file name', async () => {
       const mockFile = new File([], `bar.${ForeFlightUtils.FILE_EXTENSION}`);
       const mockContainer = ForeFlightChecklistContainer.create({ payload: { metadata: { name: 'foo' } } });
diff --git a/src/model/formats/foreflight-reader.ts b/src/model/formats/foreflight-reader.ts
index a839f2d0ca...aa21426542 100644
--- a/src/model/formats/foreflight-reader.ts
+++ b/src/model/formats/foreflight-reader.ts
@@ -19,7 +19,7 @@
 export class ForeFlightReader {
   public static async read(file: File): Promise<ChecklistFile> {
     const plaintext = await ForeFlightUtils.decrypt(await file.arrayBuffer());
-    const container: ForeFlightChecklistContainer = JSON.parse(plaintext);
+    const container: ForeFlightChecklistContainer = ForeFlightChecklistContainer.fromJsonString(plaintext);

     if (container.type !== ForeFlightUtils.CONTAINER_TYPE) {
       throw new ForeFlightFormatError(`unknown checklist type '${container.type}'`);
@@ -56,9 +57,8 @@
   }

   private static checklistGroupToEFIS(checklistGroup: ForeFlightChecklistGroup): ChecklistGroup[] {
-    const category = ForeFlightUtils.categoryToEFIS(checklistGroup.groupType);
     return checklistGroup.items
-      ? checklistGroup.items.map((item) => ForeFlightReader.checklistSubgroupToEFIS(category, item))
+      ? checklistGroup.items.map((item) => ForeFlightReader.checklistSubgroupToEFIS(checklistGroup.groupType, item))
       : []; // if the FF group (not subgroup!) is empty, drop it - we need a subgroup to represent a group
   }

diff --git a/src/model/formats/foreflight-utils.ts b/src/model/formats/foreflight-utils.ts
index 4cacacaf35...8e038cb288 100644
--- a/src/model/formats/foreflight-utils.ts
+++ b/src/model/formats/foreflight-utils.ts
@@ -1,14 +1,7 @@
-import { ChecklistGroup_Category } from '../../../gen/ts/checklist';
 import { v4 as uuid } from 'uuid';
 import { ForeFlightFormatError } from './foreflight-format';
 import { ForeFlightChecklistContainer } from '../../../gen/ts/foreflight';

-export enum ForeFlightCategory {
-  Normal = 'normal',
-  Abnormal = 'abnormal',
-  Emergency = 'emergency',
-}
-
 enum KeyUsage {
   Encrypt = 'encrypt',
   Decrypt = 'decrypt',
@@ -69,19 +62,6 @@
     );
   }

-  public static categoryToEFIS(category: string): ChecklistGroup_Category {
-    switch (category) {
-      case ForeFlightCategory.Normal:
-        return ChecklistGroup_Category.NORMAL;
-      case ForeFlightCategory.Abnormal:
-        return ChecklistGroup_Category.ABNORMAL;
-      case ForeFlightCategory.Emergency:
-        return ChecklistGroup_Category.EMERGENCY;
-      default:
-        throw new ForeFlightFormatError(`unknown category '${category}'`);
-    }
-  }
-
   public static async getObjectId(): Promise<string> {
     const hash = Array.from(
       new Uint8Array(await window.crypto.subtle.digest('SHA-256', ForeFlightUtils.encoder.encode(uuid()))),
diff --git a/src/model/formats/foreflight-writer.ts b/src/model/formats/foreflight-writer.ts
index 9bc199ce70...e2080079e8 100644
--- a/src/model/formats/foreflight-writer.ts
+++ b/src/model/formats/foreflight-writer.ts
@@ -1,11 +1,11 @@
-import { ChecklistFile, ChecklistGroup, ChecklistItem, ChecklistItem_Type } from '../../../gen/ts/checklist';
+import { ChecklistFile, ChecklistGroup, ChecklistGroup_Category, ChecklistItem, ChecklistItem_Type } from '../../../gen/ts/checklist';
 import {
   ForeFlightChecklistContainer,
   ForeFlightChecklistGroup,
   ForeFlightChecklistItem,
 } from '../../../gen/ts/foreflight';

-import { ForeFlightCategory, ForeFlightUtils } from './foreflight-utils';
+import { ForeFlightUtils } from './foreflight-utils';

 export class ForeFlightWriter {
   public static async write(file: ChecklistFile): Promise<Blob> {
@@ -32,14 +32,14 @@

   private static async checklistGroupsToFF(groupsEFIS: ChecklistGroup[]): Promise<ForeFlightChecklistGroup[]> {
     return Promise.all(
-      Object.values(ForeFlightCategory).map(
+      [ChecklistGroup_Category.NORMAL, ChecklistGroup_Category.ABNORMAL, ChecklistGroup_Category.EMERGENCY].map(
         async (categoryFF): Promise<ForeFlightChecklistGroup> => ({
           objectId: await ForeFlightUtils.getObjectId(),
           groupType: categoryFF,
           items: await Promise.all(
             groupsEFIS
               .filter(
-                (checklistGroupEFIS) => checklistGroupEFIS.category === ForeFlightUtils.categoryToEFIS(categoryFF),
+                (checklistGroupEFIS) => checklistGroupEFIS.category === categoryFF,
               )
               .map(async (checklistGroupEFIS) => ({
                 objectId: await ForeFlightUtils.getObjectId(),
diff --git a/src/model/proto/foreflight.proto b/src/model/proto/foreflight.proto
index 2ea5ca18ac...dc1f79f880 100644
--- a/src/model/proto/foreflight.proto
+++ b/src/model/proto/foreflight.proto
@@ -4,6 +4,8 @@

 option optimize_for = CODE_SIZE;

+import "checklist.proto";
+
 message ForeFlightChecklistContainer {
   string type = 1; // Always "checklist"
   ForeFlightChecklistPayload payload = 2;
@@ -25,7 +27,7 @@

 message ForeFlightChecklistGroup {
   string object_id = 1;
-  string group_type = 2;
+  efis.ChecklistGroup.Category group_type = 2;
   repeated ForeFlightChecklistSubgroup items = 3; // This key might be missing altogether
 }
```

If you see an easy solution for the case-sensitivity, great, if not, we can just leave a TODO or something to simplify it later.

Reply:
Additional patch that does make it work:

```
diff --git a/src/model/formats/foreflight-writer.ts b/src/model/formats/foreflight-writer.ts
index e2080079e8...4340c9ea8f 100644
--- a/src/model/formats/foreflight-writer.ts
+++ b/src/model/formats/foreflight-writer.ts
@@ -32,7 +32,7 @@
 
   private static async checklistGroupsToFF(groupsEFIS: ChecklistGroup[]): Promise<ForeFlightChecklistGroup[]> {
     return Promise.all(
-      [ChecklistGroup_Category.NORMAL, ChecklistGroup_Category.ABNORMAL, ChecklistGroup_Category.EMERGENCY].map(
+      [ChecklistGroup_Category.normal, ChecklistGroup_Category.abnormal, ChecklistGroup_Category.emergency].map(
         async (categoryFF): Promise<ForeFlightChecklistGroup> => ({
           objectId: await ForeFlightUtils.getObjectId(),
           groupType: categoryFF,
diff --git a/src/model/formats/test-data.ts b/src/model/formats/test-data.ts
index 43a91739ce...d37a61ffb8 100644
--- a/src/model/formats/test-data.ts
+++ b/src/model/formats/test-data.ts
@@ -87,14 +87,14 @@
 
 export const EXPECTED_FOREFLIGHT_CONTENTS = {
   groups: [
-    { category: ChecklistGroup_Category.NORMAL, title: 'Empty subgroup', checklists: [] },
+    { category: ChecklistGroup_Category.normal, title: 'Empty subgroup', checklists: [] },
     {
-      category: ChecklistGroup_Category.NORMAL,
+      category: ChecklistGroup_Category.normal,
       title: 'Subgroup with empty checklist',
       checklists: [{ title: 'Empty checklist', items: [] }],
     },
     {
-      category: ChecklistGroup_Category.NORMAL,
+      category: ChecklistGroup_Category.normal,
       title: 'Subgroup',
       checklists: [
         {
@@ -162,7 +162,7 @@
         },
       ],
     },
-    { category: ChecklistGroup_Category.ABNORMAL, title: 'Test empty abnormal subgroup', checklists: [] },
+    { category: ChecklistGroup_Category.abnormal, title: 'Test empty abnormal subgroup', checklists: [] },
   ],
   metadata: {
     name: 'Checklist name',
diff --git a/src/model/proto/checklist.proto b/src/model/proto/checklist.proto
index 6f01f0c6d1...d2ae74adff 100644
--- a/src/model/proto/checklist.proto
+++ b/src/model/proto/checklist.proto
@@ -26,11 +26,13 @@
   string title = 1;
   repeated Checklist checklists = 2;
 
+  // TODO: Make these uppercase when ts-protobuf supports customizing
+  // JSON output to match Foreflight's expectations.
   enum Category {
     UNKNOWN = 0;
-    NORMAL = 1;
-    ABNORMAL = 2;
-    EMERGENCY = 3;
+    normal = 1;
+    abnormal = 2;
+    emergency = 3;
   }
 
   Category category = 3;
```

Reply:
Thanks for the patch, this clears up a lot of things for me! I was missing one non-technical and two technical items:

1. I wasn't sure if you would accept a FF-based category enum in your model, so I decided to decouple it first.
2. I missed `fromJsonString` in the protobuf documentation.
3. Ditto for the lowercase/string enum representation trick.

I have now committed a new version in a5de561 that hopefully addresses your concerns:

1. Mapping is removed, protos are coupled.
2. "Unknown" member is removed, because if there is no mapping, I see no value in it.


I've also tried to be more consistent with the teminology.

If you don't like the magic of enum iteration, I can change it to using hardcoded keys. I'm always afraid to do things like that, as it's easier to change things in one place and forget it in another.

I left your comment about customizing the output, but I doubt that'll ever happen:

https://github.com/golang/protobuf/issues/613

The last comment is pretty clear:

> If your service really needs to produce enum values in lowercase, then perhaps define them in lowercase in the proto file.

I'd rather remove the TODO if you don't mind.

Reply:
Thanks, what you pushed LGTM.

> I'd rather remove the TODO if you don't mind.

Yes, I'm fine with that.

Reply:
Removed TODO in 8da2814.